File: PS/이것이 코딩테스트다/DFS+BFS/18_괄호변환.cpp
#include <string>
#include <vector>
#include <stack>

using namespace std;


bool isRightString(string input) {

	stack<char>s;

	for (int i = 0; i < input.length(); i++) {
		if (input[i] == '(') {
			s.push(input[i]);
		}
		else {
			if (s.empty()) {
				return false;
			}
			else {
				s.pop();
			}
		}
	}
	return true;
}

string check(string p) {
	string temp = "";
	if (p == "") {
		return "";
	}

	int count1 = 0, count2 = 0;
	string u, v;
	for (int i = 0; i < p.length(); i++) {
		if (p[i] == '(') {
			count1++;
		}
		if (p[i] == ')') {
			count2++;
		}
		if (count1 == count2) {
			u = p.substr(0, i + 1);
			v = p.substr(i + 1);
			break;
		}
	}

	if (isRightString(u)) {
		v = check(v);
		temp = u + v;
		return temp;
	}else {
		temp += "(";
		temp += check(v);
		temp += ")";
		u.erase(0,1);
		u.erase(u.length() - 1, 1);
		for (int i = 0; i < u.length(); i++) {
			if (u[i] == '(') u[i] = ')';
			else u[i] = '(';
		}
		temp += u;
		return temp;
	}
}

string solution(string p) {
	string answer = "";

	answer = check(p);
	return answer;
}

File: PS/이것이 코딩테스트다/DFS+BFS/18 괄호 변환.py
def isCorrect(p):
    stack = []
    for i in range(len(p)):
        if p[i] == "(":
            stack.append("(")
        else:
            if stack and stack[-1] == "(":
                stack.pop()
            else:
                return False
    if stack:
        return False

    return True


def balanced_index(p):
    count = 0
    for i in range(len(p)):
        if p[i] == "(":
            count += 1
        else:
            count -= 1
        if count == 0:
            return i


def solution(p):
    answer = ''
    if p == "":
        return answer
    if isCorrect(p):
        answer = p
    else:
        idx = balanced_index(p)
        u = p[:idx+1]
        v = p[idx+1:]
        if isCorrect(u):
            answer = u+solution(v)
        else:
            tmp = ""
            for i in range(1, len(u)-1):
                if u[i] == "(":
                    tmp += ")"
                else:
                    tmp += "("
            answer = "("+solution(v)+")"+tmp
    return answer
